File: libraries/mers/msg1_form_adam.f
      SUBROUTINE MSG1_FORM( PARAM, TEXT, CLEAN, MSGSTR, MSGLEN, STATUS )
*+
*  Name:
*     MSG1_FORM

*  Purpose:
*     Form a message from its text and components.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL MSG1_FORM( PARAM, TEXT, CLEAN, MSGSTR, MSGLEN, STATUS )

*  Description:
*     Construct the final text, MSGSTR( 1 : MSGLEN ), of a message
*     using the text in TEXT, and available message tokens, keywords and 
*     parameter-object associations. 
*     Clean any non-printing characters from the final string if CLEAN
*     is TRUE.

*  Arguments:
*     PARAM = CHARACTER * ( * ) (Given)
*        The message parameter name.
*     TEXT = CHARACTER * ( * ) (Given)
*        The input message text, with any tokens.
*     CLEAN = LOGICAL (Given)
*        If the string is to be 'cleaned'
*     MSGSTR = CHARACTER * ( * ) (Returned)
*        Resultant message text, with parsed tokens.
*     MSGLEN = INTEGER (Returned)
*        The filled length of MSGSTR.
*     STATUS = INTEGER (Given)
*        The global status.

*  Implementation Notes:
*     -  This subroutine is the ADAM version of MSG1_FORM.

*  Algorithm:
*     -  Attempt to get a message text from the parameter system,
*     otherwise use TEXT. 
*     -  Parse the message text and copy it into MSGSTR, making 
*     translations for parameter, status and token escapes.
*     -  The message length is returned in MSGLEN.

*  Copyright:
*     Copyright (C) 1982, 1984, 1985, 1989, 1990, 1991, 1994 Science & Engineering Research Council.
*     Copyright (C) 1999, 2001 Central Laboratory of the Research Councils.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*     
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*     
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 59 Temple Place,Suite 330, Boston, MA
*     02111-1307, USA

*  Authors:
*     JRG: Jack Giddings (UCL)
*     BDK: Dennis Kelly (ROE)
*     AJC: Alan Chipperfield (STARLINK)
*     PCTR: P.C.T. Rees (STARLINK)
*     BKM: B.K.McIlwrath (STARLINK)
*     {enter_new_authors_here}

*  History:
*     3-JAN-1982 (JRG):
*        Original version.
*     13-NOV-1984 (BDK):
*        ADAM version.
*     11-JUN-1985 (BDK):
*        Discard leading % from error message.
*     5-JUN-1989 (AJC):
*        Check whole string for equality with escape.
*     13-SEP_1989 (PCTR):
*        Converted to new prologue and layout.
*     13-MAR-1990 (PCTR):
*        Converted to use EMS_ calls and changed subroutine name.
*     9-APR-1990 (PCTR):
*        Replaced DO WHILE construct with ANSI Fortran 77 equivalent.
*     9-OCT-1991 (PCTR):
*        New token parsing algorithm.
*     24-OCT-1991 (PCTR):
*        Removed bugs from new token parsing algorithm.
*     12-AUG-1994 (BKM):
*        Change reserved token STATUS to call EMS_FACER (from EMS_SYSER)
*     15-SEP-1999 (AJC):
*        Add argument CLEAN
*     21-FEB-2001 (AJC):
*        Remove use of internal EMS routines
*        Use EMS_EXPND instead of EMS1_GTOK
*        EMS1_PUTC is now MSG1_PUTC
*        Use MSG1_KTOK not EMS1_KTOK
*     19-OCT-2001 (AJC):
*        Correct finding ^STATUS if previous token in message
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type Definitions:
      IMPLICIT NONE                     ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'                 ! Standard SAE constants
      INCLUDE 'EMS_PAR'                 ! EMS_ public constants
      INCLUDE 'MSG_PAR'                 ! MSG_ public constants
      INCLUDE 'MSG_SYS'                 ! MSG_ private constants

*  Arguments Given:
      CHARACTER * ( * ) PARAM
      CHARACTER * ( * ) TEXT
      LOGICAL CLEAN

*  Arguments Returned:
      CHARACTER * ( * ) MSGSTR

      INTEGER MSGLEN

*  Status:
      INTEGER STATUS

*  External References:
      LOGICAL MSG1_GENV                 ! Get the message text
      LOGICAL MSG1_GKEY                 ! Get the parameter keyword
      LOGICAL MSG1_GREF                 ! Get the parameter reference

      LOGICAL CHR_SIMLR                 ! Caseless string compariso
      INTEGER CHR_LEN                   ! Character string length

*  Local Variables:
      LOGICAL FOUND                     ! ^STATUS found
      LOGICAL DEFINE                    ! Whether a token is defined
      LOGICAL DOUBLE                    ! Whether a double escape
      LOGICAL LITERL                    ! Whether a literal token escape

      INTEGER ITOK                      ! Position of STATUS token
      INTEGER STMLEN                    ! Length of STATUS message
      INTEGER CURPOS                    ! Current character position
      INTEGER LSTAT                     ! Local status
      INTEGER LSTPOS                    ! Previous value of CURPOS
      INTEGER NAMLEN                    ! Name string length
      INTEGER PSTAT                     ! Local status
      INTEGER TEXLEN                    ! Message text length
      INTEGER TKVLEN                    ! Token value length

      CHARACTER * 1 ESCAPE              ! Current escape character
      CHARACTER * ( 2 ) ESCSTR          ! Escape string
      CHARACTER * ( MSG__SZMSG ) NAMSTR ! Token name string
      CHARACTER * ( MSG__SZTOK ) STMSG  ! STATUS message
      CHARACTER * 1 PREVEC              ! Previous escape character
      CHARACTER * ( MSG__SZMSG ) TEXST1 ! Message text string
      CHARACTER * ( MSG__SZMSG ) TEXST2 ! Part expanded text string
      CHARACTER * ( MSG__SZTOK ) TOKVAL ! Token value string

*.

*  Operate regardless of STATUS - use local status
      LSTAT = SAI__OK

*  Select the message text from the interface module or program.
      IF ( .NOT. MSG1_GENV( PARAM, TEXST1, TEXLEN ) ) THEN
         TEXST1 = TEXT
         TEXLEN = CHR_LEN( TEXST1 )
      END IF

*  Initialise the returned message.

      MSGSTR = ' '
      MSGLEN = 1

*  Check for an empty string.
      IF ( TEXLEN .GT. 0 ) THEN

*     Check for the STATUS token in text
*     If it's there, substitute it here and now.
*     We must do this to avoid concatenation if ^STATUS is used after RENEW.
*     However, it means that ^STATUS is now truly reserved and cannot be
*     preset for concatenation.
*     We assume there is not more than one occurrence of ^STATUS within TEXT
         CURPOS = 1
         FOUND = .FALSE.

*     Find a token
         ITOK = INDEX( TEXST1(1:TEXLEN), '^' )
*        DOWHILE loop
 10      CONTINUE

         IF ( .NOT. FOUND
     :        .AND. ( ITOK .GT. 0 )
     :        .AND. ( ITOK .LT. TEXLEN ) ) THEN

*     Is it STATUS
            NAMSTR = TEXST1(ITOK+1:ITOK+6)            
            IF( CHR_SIMLR( NAMSTR, 'STATUS' ) ) THEN
*       It is STATUS
*       Find the associated message
               CALL EMS_MARK
               CALL EMS_FACER( 'STATUS', STATUS )
               CALL EMS_EXPND( '^STATUS', STMSG, TKVLEN, LSTAT )

*       Save the remainder of TEXST1
               IF ( TEXLEN .GT. ITOK + 7 ) 
     :           TEXST2 = TEXST1(ITOK+7:TEXLEN)

*       Copy the message to the ^STATUS position
               CURPOS = ITOK - 1
               CALL MSG1_PUTC( STMSG(1:TKVLEN), TEXST1, CURPOS, LSTAT )
*       and add any remaining part of message
               IF ( TEXLEN .GT. ITOK + 7 ) 
     :           CALL MSG1_PUTC(
     :             TEXST2(1:TEXLEN-(ITOK+7)+1), TEXST1, CURPOS, LSTAT )
               TEXLEN = CURPOS
               CALL EMS_RLSE
*     We assume there's only one ^STATUS
               FOUND = .TRUE.

            END IF

            LSTPOS = ITOK
            ITOK = INDEX( TEXST1(LSTPOS+1:TEXLEN), '^' )
            IF ( ITOK .GT. 0 ) ITOK = LSTPOS + ITOK

*     Continue loop
            GO TO 10
         END IF

*     Make a first pass to expand plain ^ tokens
*     This will also kill tokens
         CALL EMS_EXPND(
     :     TEXST1(1:TEXLEN), TEXST2, TEXLEN, LSTAT )

*     Initialise the returned message length.
         MSGLEN = 0

*     Initialise token escape state flag.
         LITERL = .FALSE.

*     Initialise the text pointers and local status.
         CURPOS = 0
         LSTPOS = 0
         LSTAT = SAI__OK
         PSTAT = SAI__OK

*     Initialise the escape string and previous escape character.
         ESCSTR = MSG__KEYEC // MSG__REFEC
         PREVEC = ' '
         ESCAPE = ' '

*     Parse and translate the returned message text.
*     DO WHILE loop.
 20      CONTINUE
         IF ( PSTAT .EQ. SAI__OK .AND. CURPOS .LT. TEXLEN ) THEN 

*        Check if a double or paired escape character sequence has 
*        occurred.
            IF ( ESCAPE .EQ. PREVEC ) THEN
               DOUBLE = .TRUE.
            ELSE
               DOUBLE = .FALSE.
            END IF

         ELSE

*        No paired token escapes, so annul the DOUBLE flag
            DOUBLE = .FALSE.

         END IF

*     Find the next occurrence of an escape character.
         CALL MSG1_GESC( ESCSTR, TEXST2( 1 : TEXLEN ), CURPOS )

*     Append any text prior to the escape character to the returned 
*     string.
         IF ( CURPOS .EQ. 0 ) THEN

*        No more escape characters have been found, so append all 
*        the text that remains to the returned message text and exit 
*        the loop.
            CALL MSG1_PUTC( TEXST2( LSTPOS+1 : TEXLEN ), MSGSTR, 
     :                      MSGLEN, PSTAT )
            GO TO 30

         ELSE
*        A token escape has been found, so get which escape character 
*        has occurred.
            ESCAPE = TEXST2( CURPOS : CURPOS )

*        Check if it is a double token escape.
            IF ( LITERL .AND. CURPOS .EQ. LSTPOS+1 ) THEN

*           Check if a double escape character sequence has occurred.
               IF ( ESCAPE .EQ. PREVEC ) THEN
                  DOUBLE = .TRUE.
               ELSE
                  DOUBLE = .FALSE.
               END IF
            ELSE

*           No paired token escapes, so annul the DOUBLE flag
               DOUBLE = .FALSE.
            END IF

*           Act.
            IF ( DOUBLE ) THEN

*           A double token escape, so do nothing except update
*              CURPOS and reset the literal token escape flag.
               LITERL = .FALSE.

*           Assign a "null" ESCAPE.
               ESCAPE = ' '

            ELSE
*           Reset the literal token escape flag.
               LITERL = .FALSE.

*           Append any text prior to the escape character.
               IF ( CURPOS .GT. 1 ) CALL MSG1_PUTC( 
     :                                 TEXST2( LSTPOS+1 : CURPOS-1 ), 
     :                                 MSGSTR, MSGLEN, PSTAT )

*           Find the token name.
               CALL MSG1_GNAM( TEXST2( 1 : TEXLEN ), CURPOS, NAMSTR,
     :                            NAMLEN, LSTAT )

*           Check that a token name exists.
               IF ( LSTAT .NE. SAI__OK ) THEN

*              The name string has been over-run, so indicate this 
*              in the message text.
                  CALL MSG1_PUTC( ESCAPE, MSGSTR, MSGLEN, PSTAT )
                  CALL MSG1_PUTC( '<', MSGSTR, MSGLEN, PSTAT )
                  CALL MSG1_PUTC( NAMSTR( 1 : NAMLEN ), MSGSTR, 
     :                               MSGLEN, PSTAT )
                  CALL MSG1_PUTC( '>', MSGSTR, MSGLEN, PSTAT )

*              Reset the local status.
                  LSTAT = SAI__OK

               ELSE IF ( NAMLEN .GT. 0 ) THEN

                  IF ( ESCAPE .EQ. MSG__REFEC ) THEN

*                 Get a value for the object name token.
                     DEFINE = MSG1_GREF( NAMSTR( 1 : NAMLEN ), 
     :                                      TOKVAL, TKVLEN )

                  ELSE IF ( ESCAPE .EQ. MSG__KEYEC ) THEN

*                 Get a value for the keyword token.
                     DEFINE = MSG1_GKEY( NAMSTR( 1 : NAMLEN ), 
     :                                      TOKVAL, TKVLEN )
                  END IF

*              Check if the token value was defined.
                  IF ( DEFINE ) THEN

*                 Append the token value to the returned string.
                     CALL MSG1_PUTC( TOKVAL( 1 : TKVLEN ), MSGSTR, 
     :                                  MSGLEN, PSTAT )
                  ELSE

*                 A value has not been found, so append the 
*                 "undefined" token string.
                     CALL MSG1_PUTC( ESCAPE, MSGSTR, MSGLEN, PSTAT )
                     CALL MSG1_PUTC( '<', MSGSTR, MSGLEN, PSTAT )
                     CALL MSG1_PUTC( NAMSTR( 1 : NAMLEN ), MSGSTR, 
     :                                  MSGLEN, PSTAT )
                     CALL MSG1_PUTC( '>', MSGSTR, MSGLEN, PSTAT )
                  END IF

*              Assign a "null" ESCAPE.
                  ESCAPE = ' '

               ELSE
*              There is an isolated token escape character in the 
*              string, so set the literal token escape flag.
                  LITERL = .TRUE.
                  CALL MSG1_PUTC( ESCAPE, MSGSTR, MSGLEN, PSTAT )

               END IF
            END IF

*        Update LSTPOS and PREVEC.
            LSTPOS = CURPOS
            PREVEC = ESCAPE
            GO TO 20
         END IF

 30   CONTINUE
      END IF

*  Clean the returned message string unless in STREAM mode.
      IF ( CLEAN ) CALL CHR_CLEAN( MSGSTR( 1 : MSGLEN ) )

*  Clear the message token table.
!      CALL MSG1_KTOK
 
      END

File: libraries/mers/msg1Form_adam.c
/*
*+
*  Name:
*     msg1Form

*  Purpose:
*     Form a message from its text and components.

*  Language:
*     Starlink ANSI C

*  Invocation:
*     msg1Form( const char *param, const char *text, int clean,
*               int useformat, size_t msglen, char *msgstr, int * status );

*  Description:
*     Construct the final text of a message using the text in "text",
*     and available message tokens, keywords and parameter-object associations.
*     Clean any non-printing characters from the final string if "clean"
*     is true.
*
*     Keyword associations are disabled if "useformat" is true.

*  Arguments:
*     param = const char * (Given)
*        The message parameter name.
*     text = const char * (Given)
*        The input message text, with any tokens.
*     clean = int (Given)
*        If the string is to be 'cleaned'
*     useformat = int (Given)
*        Boolean indicating whether the supplied text needs to take account
*        of sprintf formatting. See notes for additional information.
*     msglen = size_t (Given)
*        Allocated size of buffer "msgstr".
*     msgstr = char * (Returned)
*        Resultant message text, with parsed tokens.
*     status = int * (Given)
*        The global status.

*  Implementation Notes:
*     -  This subroutine is the ADAM version of msg1Form

*  Notes:
*     If "useformat" is true Keyword association is disabled because the
*     "%KEYWORD" syntax is hard to disambiguate from a sprintf %x specifier.
*     Additionally, token replacement will be modified to replace any %
*     in tokens with %% in preparation for a sprintf call.

*  Algorithm:
*     -  Attempt to get a message text from the parameter system,
*     otherwise use "text".
*     -  Parse the message text and copy it into "msgstr", making
*     translations for parameter, status and token escapes.
*     -  The string will be null terminated.

*  Copyright:
*     Copyright (C) 1982, 1984, 1985, 1989-1991, 1994 Science & Engineering
*     Research Council.
*     Copyright (C) 1999, 2001 Central Laboratory of the Research Councils.
*     Copyright (C) 2008 Science and Technology Facilities Council.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     JRG: Jack Giddings (UCL)
*     BDK: Dennis Kelly (ROE)
*     AJC: Alan Chipperfield (STARLINK)
*     PCTR: P.C.T. Rees (STARLINK)
*     BKM: B.K.McIlwrath (STARLINK)
*     TIMJ: Tim Jenness (JAC, Hawaii)
*     {enter_new_authors_here}

*  History:
*     3-JAN-1982 (JRG):
*        Original version.
*     13-NOV-1984 (BDK):
*        ADAM version.
*     11-JUN-1985 (BDK):
*        Discard leading % from error message.
*     5-JUN-1989 (AJC):
*        Check whole string for equality with escape.
*     13-SEP_1989 (PCTR):
*        Converted to new prologue and layout.
*     13-MAR-1990 (PCTR):
*        Converted to use EMS_ calls and changed subroutine name.
*     9-APR-1990 (PCTR):
*        Replaced DO WHILE construct with ANSI Fortran 77 equivalent.
*     9-OCT-1991 (PCTR):
*        New token parsing algorithm.
*     24-OCT-1991 (PCTR):
*        Removed bugs from new token parsing algorithm.
*     12-AUG-1994 (BKM):
*        Change reserved token STATUS to call EMS_FACER (from EMS_SYSER)
*     15-SEP-1999 (AJC):
*        Add argument CLEAN
*     21-FEB-2001 (AJC):
*        Remove use of internal EMS routines
*        Use EMS_EXPND instead of EMS1_GTOK
*        EMS1_PUTC is now MSG1_PUTC
*        Use MSG1_KTOK not EMS1_KTOK
*     19-OCT-2001 (AJC):
*        Correct finding ^STATUS if previous token in message
*     05-SEP-2008 (TIMJ):
*        Rewrite in C.
*     26-NOV-2008 (TIMJ):
*        Fix ^STATUS expansion.
*     23-DEC-2008 (TIMJ):
*        Disable keyword associations if sprintf formats are required.
*     12-JAN-2008 (TIMJ):
*        Fix logic with STREAM mode and with % escaping.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-
*/

/* for strtok_r in standards compliance mode */
#if HAVE_STRTOK_R
#define _POSIX_C_SOURCE 200112L
#endif

#if HAVE_CONFIG_H
#  include <config.h>
#endif

#include "sae_par.h"
#include "ems.h"
#include "msg_par.h"
#include "msg_sys.h"

#include "star/util.h"

#include "mers1.h"

#include <ctype.h>
#include <string.h>

/* Name of token representing STATUS error string */
#define STATSTR "STATUS"

void msg1Form ( const char * param, const char * text, int clean,
                int useformat, size_t msglen, char * msgstr, int * status ) {

  /*  Local Variables: */
  int define;          /* Whether a token is defined */
  int hasdouble = 0;   /* Whether a double escape */
  int literl;          /* Whether a literal token escape */

  int iposn;           /* Position in output string */
  int curpos;          /* Current character position */
  int lstat = SAI__OK; /* Local status */
  int lstpos;          /* Previous value of curpos */
  int namlen;          /* Name string length */
  int oldstm;          /* Initial EMS STREAM setting */
  int pstat = SAI__OK; /* Local status */
  int texlen;          /* message text length */
  int tkvlen;          /* Token value length */

  char escape[2];      /* Current escape character */
  char escstr[3];      /* Escape string */
  char namstr[MSG__SZMSG]; /* Token name string */
  char stmsg[MSG__SZTOK];  /* STATUS message */
  char prevec[2];      /* Previous escape character */
  char texst0[MSG__SZMSG];
  char texst1[MSG__SZMSG]; /* Message text string */
  char texst2[MSG__SZMSG]; /* Part expanded text string */
  char tokval[MSG__SZTOK]; /* Token value string */

  char *tokctx = NULL;   /* Context argument for strtok_r */
  char *token = NULL;    /* Pointer to token in string */

  size_t szst;           /* Length of STATUS string */
  ptrdiff_t offset;      /* offset into string */

  /* Can not do anything if we have a null pointer */
  if (!msgstr || msglen == 0) return;

  /*  Operate regardless of STATUS - use local status */
  lstat = SAI__OK;

  /*  Select the message text from the interface module or program. */
  if ( !msg1Genv( param, texst1, sizeof(texst1) ) ) {
    star_strlcpy( texst1, text, sizeof(texst1) );
  }

  /*  Initialise the returned message. */
  msgstr[0] = '\0';

  /*  Check for an empty string. */
  if (strlen(texst1)) {

    /* Check for the STATUS token in text If it's there, substitute it
     * here and now.  We must do this to avoid concatenation if
     * ^STATUS is used after RENEW.  However, it means that ^STATUS is
     * now truly reserved and cannot be preset for concatenation.  We
     * assume there is not more than one occurrence of ^STATUS within
     * TEXT */

    /* Since sttrok twiddles with the string we need to take a copy */
    star_strlcpy( texst0, texst1, sizeof(texst0) );

    /* Find a token */
#if HAVE_STRTOK_R
    token = strtok_r( texst1, "^", &tokctx );
#else
    token = strtok( texst1, "^" );
#endif

    /* Get the length of the token string we are looking for */
    szst = strlen( STATSTR );

    while (token) {

      /* Is it STATUS?  - noting that ^STATUSX will also match.
       Note that first time round the loop strtok will return the start
       of string so we have to check that we do not match a bare STATUS
       at the start (there can never be a token as first character since
       a ^ would point us to the second character). */
      if (token != texst1 && strncmp( token, STATSTR, szst ) == 0) {

        /*       It is STATUS
         *       Find the associated message */
        emsMark();
        emsFacer( STATSTR, *status );
        emsExpnd( "^" STATSTR, stmsg, sizeof(stmsg), useformat, &tkvlen,
                  &lstat );
        emsRlse();

        /* Need to put this text back into original */

        /* Save the text that follows ^STATUS - it has to be copied
         from texst0 since token aka texst1 is modified by strok on the
         first call. */
        offset = token - texst1 + szst;
        star_strlcpy( texst2, &(texst0[offset]), sizeof(texst2) );

        /* Work out how far into the buffer this ^STATUS was found, accounting
         for the fact that "token" reports a value one larger than the position
         of the ^ */
        offset = token - texst1 - 1;

        /* Append the text from the STATUS token and the subsequent text */
        texst0[offset] = '\0';
        star_strappend( texst0, stmsg, sizeof(texst0) );
        star_strappend( texst0, texst2, sizeof(texst0) );

        break;

      }

      /* Keep on looking */
#if HAVE_STRTOK_R
      token = strtok_r( NULL, "^", &tokctx );
#else
      token = strtok_r( NULL, "^" );
#endif
    }

    /*     Make a first pass to expand plain ^ tokens
     *     This will also kill tokens  - note that we have to force
     *     STREAM mode in EMS so that we can control it ourselves
     *     later on */
    oldstm = emsStune( "STREAM", !clean, &lstat );
    emsExpnd( texst0, texst2, sizeof(texst2), useformat, &texlen, &lstat);
    (void) emsStune( "STREAM", oldstm, &lstat );

    /*     Initialise token escape state flag. */
    literl = 0;

    /*     Initialise the text pointers and local status. */
    curpos = -1;
    lstpos = -1;
    iposn = 0;
    lstat = SAI__OK;
    pstat = SAI__OK;

    /*     Initialise the escape string and previous escape character.
           Note that KEY escapes are not included if useformat is true and
           if it happens to be a "%" that we use for params.
     */
    star_strlcpy( escstr, MSG__REFEC, sizeof(escstr) );
    if (!(useformat && strcmp(MSG__KEYEC, "%") == 0)) {
      star_strlcat( escstr, MSG__KEYEC, sizeof(escstr) );
    }
    prevec[0] = '\0';
    escape[0] = '\0';
    prevec[1] = '\0';
    escape[1] = '\0';

    /*     Parse and translate the returned message text. */
    while (1) {

      if (pstat == SAI__OK && curpos < texlen) {

        /*        Check if a double or paired escape character sequence has
         *        occurred. */
        if (prevec[0] == escape[0] ) {
          hasdouble = 1;
        } else {
          hasdouble = 0;
        }

      } else {

        /*        No paired token escapes, so annul the DOUBLE flag */
        hasdouble = 0;

      }

      /*     Find the next occurrence of an escape character. */
      ems1Gesc( escstr, texst2, &curpos );

      /*     Append any text prior to the escape character to the returned
       *     string. */
      if (curpos == -1) {

        /*        No more escape characters have been found, so append all
         *        the text that remains to the returned message text and exit
         *        the loop. */
        pstat = star_strappend( msgstr, &(texst2[lstpos+1]), msglen );
        break;
      } else {

        /*        A token escape has been found, so get which escape character
         *        has occurred. */
        escape[0] = texst2[curpos];

        /*        Check if it is a double token escape. */
        if (literl && curpos == lstpos + 1 ) {

          /*    Check if a double escape character sequence has occurred. */
          if (escape[0] == prevec[0]) {
            hasdouble = 1;
          } else {
            hasdouble = 0;
          }

        } else {

          /*  No paired token escapes, so annul the DOUBLE flag */
          hasdouble = 0;
        }

        /*           Act. */
        if (hasdouble) {

          /*           A double token escape, so do nothing except update
           *              CURPOS and reset the literal token escape flag. */
          literl = 0;

          /*           Assign a "null" ESCAPE. */
          escape[0] = '\0';

        } else {

          /*           Reset the literal token escape flag. */
          literl = 0;

          /*           Append any text prior to the escape character. */
          if (curpos >= 0) {
            texst2[curpos] = '\0';
            star_strappend( msgstr, &texst2[lstpos+1], msglen);
            texst2[curpos] = escape[0];
          }

          /*           Find the token name - curpos is updated. */
          ems1Gnam( texst2, &curpos, namstr, &namlen, &lstat );

          /*           Check that a token name exists. */
          if (lstat != SAI__OK) {
            /*              The name string has been over-run, so indicate this
             *              in the message text. */
            star_strappend( msgstr, escape, msglen);
            star_strappend( msgstr, "<", msglen);
            star_strappend( msgstr, namstr, msglen);
            star_strappend( msgstr, ">", msglen);

            /*              Reset the local status. */
            lstat = SAI__OK;

          } else if (namlen > 0) {
            define = 0;
            if ( strcmp(escape, MSG__REFEC) == 0) {

              /*                 Get a value for the object name token. */
              define = msg1Gref( namstr, tokval, sizeof(tokval) );

            } else if (strcmp(escape, MSG__KEYEC) == 0) {

              /*                 Get a value for the keyword token.*/
              define = msg1Gkey( namstr, tokval, sizeof(tokval) );
            }

            /*          Check if the token value was defined. */
            if (define) {
              /*            Append the token value to the returned string. */
              star_strappend( msgstr, tokval, msglen );
            } else {
              /*                 A value has not been found, so append the
               *                 "undefined" token string. */
              star_strappend( msgstr, escape, msglen);
              star_strappend( msgstr, "<", msglen);
              star_strappend( msgstr, namstr, msglen);
              star_strappend( msgstr, ">", msglen);
            }

            /*              Assign a "null" ESCAPE. */
            escape[0] = '\0';

          } else {
            /*              There is an isolated token escape character in the
             *              string, so set the literal token escape flag. */
            literl = 1;
            star_strappend( msgstr, escape, msglen );

          }
        }

        /*        Update LSTPOS and PREVEC. */
        lstpos = curpos;
        prevec[0] = escape[0];
      }

    }

    /*  Clean the returned message string unless in STREAM mode. */
    if (clean) {
      size_t i;
      size_t len;
      len = strlen(msgstr);
      for (i = 0; i < len; i++) {
        if (!isprint( msgstr[i] )) {
          msgstr[i] = ' ';
        }
      }
    }

  }
}
